File: src/flutter/dart/lib/ViewFactory.dart
/*
  This file is part of KDDockWidgets.

  SPDX-FileCopyrightText: 2019-2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
  Author: Sérgio Martins <sergio.martins@kdab.com>

  SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only

  Contact KDAB at <info@kdab.com> for commercial licensing options.
*/

import 'dart:ffi';

import 'package:KDDockWidgets/FloatingWindow.dart';
import 'package:KDDockWidgets/View_mixin.dart';
import 'package:KDDockWidgetsBindings/Bindings.dart' as KDDockWidgetBindings;
import 'package:KDDockWidgetsBindings/Bindings_KDDWBindingsCore.dart'
    as KDDWBindingsCore;
import 'package:KDDockWidgetsBindings/Bindings_KDDWBindingsFlutter.dart'
    as KDDWBindingsFlutter;
import 'DockWidget.dart';
import 'DropArea.dart';
import 'Group.dart';
import 'IndicatorWindow.dart';
import 'RubberBand.dart';
import 'Separator.dart';
import 'Stack.dart';
import 'TitleBar.dart';
import 'TabBar.dart';

class ViewFactory extends KDDWBindingsFlutter.ViewFactory {
  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createDropArea(
      KDDWBindingsCore.DropArea? controller, KDDWBindingsCore.View? parent) {
    return DropArea(
        controller, KDDockWidgetBindings.Core_ViewType.DropArea, parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsFlutter.IndicatorWindow createClassicIndicatorWindow_flutter(
      KDDWBindingsCore.ClassicDropIndicatorOverlay? indicators,
      {required KDDWBindingsCore.View? parent}) {
    // Not using parent. It won't be parented on DropArea, but will be
    // a transparent top-level window
    return IndicatorWindow(indicators, null);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createGroup(KDDWBindingsCore.Group? group,
      {required KDDWBindingsCore.View? parent}) {
    return Group(group, parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createTabBar(KDDWBindingsCore.TabBar? tabBar,
      {required KDDWBindingsCore.View? parent}) {
    return TabBar(tabBar, parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createTitleBar(
      KDDWBindingsCore.TitleBar? controller, KDDWBindingsCore.View? parent) {
    return TitleBar(controller, parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createStack(
      KDDWBindingsCore.Stack? stack, KDDWBindingsCore.View? parent) {
    return Stack(stack, parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createSeparator(KDDWBindingsCore.Separator? separator,
      {required KDDWBindingsCore.View? parent}) {
    return Separator(separator, parent: parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createRubberBand(KDDWBindingsCore.View? parent) {
    return RubberBand(parent);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createDockWidget(String? uniqueName,
      {int options = 0, int layoutSaverOptions = 0, int windowFlags = 0}) {
    return DockWidget(uniqueName,
        options: options, layoutSaverOptions: layoutSaverOptions);
  }

  @override
  @pragma("vm:entry-point")
  KDDWBindingsCore.View createFloatingWindow(
      KDDWBindingsCore.FloatingWindow? controller,
      {required KDDWBindingsCore.MainWindow? parent,
      int windowFlags = 0}) {
    // TODOm4: What to do with main window pointer
    return FloatingWindow(controller, null);
  }
}

File: src/flutter/ViewFactory.cpp
/*
  This file is part of KDDockWidgets.

  SPDX-FileCopyrightText: 2019-2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
  Author: Sérgio Martins <sergio.martins@kdab.com>

  SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only

  Contact KDAB at <info@kdab.com> for commercial licensing options.
*/

#include "ViewFactory.h"
#include "Config.h"

#include "ClassicIndicatorsWindow.h"
#include "core/Utils_p.h"
#include "core/Logging_p.h"

#include "kddockwidgets/core/TabBar.h"
#include "kddockwidgets/core/Stack.h"
#include "kddockwidgets/core/FloatingWindow.h"
#include "kddockwidgets/core/indicators/ClassicDropIndicatorOverlay.h"
#include "kddockwidgets/core/indicators/NullDropIndicatorOverlay.h"
#include "kddockwidgets/core/indicators/SegmentedDropIndicatorOverlay.h"
#include "kddockwidgets/core/MainWindow.h"

// #include "qtwidgets/views/ClassicIndicatorsWindow.h"
// #include "qtwidgets/views/SegmentedDropIndicatorOverlay.h"
// #include "qtwidgets/views/FloatingWindow.h"
// #include "qtwidgets/views/DockWidget.h"
// #include "qtwidgets/views/DropArea.h"
// #include "qtwidgets/views/Group.h"
// #include "qtwidgets/views/View.h"
// #include "qtwidgets/views/Separator.h"
// #include "qtwidgets/views/TitleBar.h"
// #include "qtwidgets/views/TabBar.h"
// #include "qtwidgets/views/SideBar.h"
// #include "qtwidgets/views/Stack.h"
// #include "qtwidgets/views/MainWindow.h"
// #include "qtwidgets/views/MDILayout.h"
// #include "qtwidgets/views/RubberBand.h"

#include "views/ClassicIndicatorWindowViewInterface.h"


// clazy:excludeall=ctor-missing-parent-argument

using namespace KDDockWidgets;
using namespace KDDockWidgets::flutter;

ViewFactory::~ViewFactory()
{
}

Core::View *ViewFactory::createDockWidget(const QString &, DockWidgetOptions, LayoutSaverOptions,
                                          Qt::WindowFlags) const
{
    return {};
}


Core::View *ViewFactory::createGroup(Core::Group *, Core::View *) const
{
    Q_ASSERT(false);
    return {};
}

Core::View *ViewFactory::createTitleBar(Core::TitleBar *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory::createTabBar(Core::TabBar *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory::createStack(Core::Stack *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory::createSeparator(Core::Separator *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory::createFloatingWindow(Core::FloatingWindow *,
                                              Core::MainWindow *, Qt::WindowFlags) const
{
    return {};
}

Core::View *ViewFactory::createRubberBand(Core::View *) const
{
    return nullptr;
}

Core::View *ViewFactory::createSideBar(Core::SideBar *, Core::View *) const
{
    return {};
}

QAbstractButton *ViewFactory::createTitleBarButton(QWidget *, TitleBarButtonType) const
{
    return nullptr;
}

// iconForButtonType impl is the same for QtQuick and QtWidgets
Icon ViewFactory::iconForButtonType(TitleBarButtonType, qreal) const
{
    return {};
}

Core::View *ViewFactory::createDropArea(Core::DropArea *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory::createMDILayout(Core::MDILayout *, Core::View *) const
{
    return {};
}

Core::View *
ViewFactory::createSegmentedDropIndicatorOverlayView(Core::SegmentedDropIndicatorOverlay *,
                                                     Core::View *) const
{
    return {};
}

Core::ClassicIndicatorWindowViewInterface *
ViewFactory::createClassicIndicatorWindow(Core::ClassicDropIndicatorOverlay *controller, Core::View *parent) const
{
    // We need a little indirection here, since our bindings don't support multiple inheritance
    // In dart View factory can't return something of type ClassicIndicatorWindowViewInterface when only having IndicatorWindow
    // But in C++ we can, so get it from flutter and cast it here
    return createClassicIndicatorWindow_flutter(controller, parent);
}

flutter::IndicatorWindow *
ViewFactory::createClassicIndicatorWindow_flutter(Core::ClassicDropIndicatorOverlay *, Core::View *) const
{
    spdlog::warn("Implemented in dart {}", Q_FUNC_INFO);
    return nullptr;
}
